File: web/src/model/registry/elements/fullAdder.ts
/**
 * The ripple-carry full adder (FullAdderElm.java, dump 196): `bits` A inputs
 * and `bits` B inputs summed with a carry-in into `bits` S outputs and a
 * carry-out. The width is the `bits` file token under FLAG_BITS, or 1 for a
 * flagless line, exactly the file constructor's default (FullAdderElm.java:
 * 30-35). A fresh part carries the flag like upstream's interactive
 * constructor; a flagless file line keeps its byte-exact form on save and an
 * edit to its width sets the flag, so the next save writes the `bits` token.
 */

import {
  CHIP_BIT_ORDER_BUS,
  chipBitOrderFlags,
  chipBitOrderParam,
  chipBodyRect,
  chipCommonTokens,
  chipDumpFlags,
  chipPosts,
  drawChip,
  normalizeChipBits,
  type ChipPinDef,
} from './dFlipFlop';
import type { CircuitElement, DrawContext, ElementDef } from '../../types';

/** File flag saying a `bits` token follows the common chip fields
 *  (FullAdderElm.java:36). */
export const FULL_ADDER_BITS = 2;

/** The bit width, truncated and clamped to the 1..16 the engine accepts
 *  (full_adder.rs:38). */
export function normalizeFullAdderBits(value: number): number {
  return normalizeChipBits(value, 1, 16);
}

function fullAdderBits(e: CircuitElement): number {
  if ((e.flags & FULL_ADDER_BITS) !== 0) return normalizeFullAdderBits(e.params.bits ?? 4);
  return 1;
}

/** Bus bit order (upstream BIT_ORDER_BUS, XML attribute `bo="2"`): the A, B
 *  and S groups each collapse onto one row of their side. Upstream only
 *  allows it under FLAG_BITS (allowBus = needsBits). */
export function fullAdderBus(e: CircuitElement): boolean {
  return (
    e.params.bitOrder === 2 ||
    ((e.flags & (CHIP_BIT_ORDER_BUS | FULL_ADDER_BITS)) === (CHIP_BIT_ORDER_BUS | FULL_ADDER_BITS))
  );
}

function fullAdderSizeY(e: CircuitElement): number {
  // bitsY is 1 in bus mode and `bits` otherwise (FullAdderElm.java:42-43).
  return fullAdderBus(e) ? 3 : 2 * fullAdderBits(e) + 1;
}

/** The pin table, from `setupPins` (FullAdderElm.java:41-56). Non-bus, the
 *  MSB-first default lays each group out high row first (makeBitPins places
 *  name_i at row pos+count-1-i), so A3..A0 descend the west and the S outputs
 *  start two rows inside the east edge. In bus mode each group collapses onto
 *  its anchor row carrying per-bit tags (the useBus branch), which is what
 *  keeps td4's adder wired to its bus wires. */
export function fullAdderPins(e: CircuitElement): ChipPinDef[] {
  const bits = fullAdderBits(e);
  const bus = fullAdderBus(e);
  const pins: ChipPinDef[] = [];
  const bank = (side: 'W' | 'E', pos: number, name: string, output: boolean): void => {
    for (let i = 0; i < bits; i++) {
      pins.push(
        bus
          ? { side, pos, text: name, output, busWidth: bits, busZ: i }
          : { side, pos: pos + (bits - 1 - i), text: `${name}${i}`, output },
      );
    }
  };
  bank('W', 0, 'A', false);
  // The B group's anchor row is bitsY, `bits` outside bus mode.
  bank('W', bus ? 1 : bits, 'B', false);
  bank('E', 2, 'S', true);
  // The carry-in sits at bitsY*2 (FullAdderElm.java:54).
  pins.push({ side: 'W', pos: bus ? 2 : 2 * bits, text: 'Cin' });
  pins.push({ side: 'E', pos: 0, text: 'C', output: true });
  return pins;
}

function drawFullAdder(g: DrawContext, e: CircuitElement): void {
  drawChip(g, e, 2, fullAdderSizeY(e), fullAdderPins(e));
}

export const FULL_ADDER_DEF: ElementDef = {
  kind: 'fullAdder',
  label: 'adder',
  category: 'Logic',
  dumpCode: '196',
  postCount: 14,  // the default 4-bit layout: 3*4 + the carry pair
  posts: (e) => chipPosts(e, 2, fullAdderSizeY(e), fullAdderPins(e)),
  bodyRect: (e) => chipBodyRect(e, 2, fullAdderSizeY(e)),
  noDiagonal: true,  // ChipElm.java:44
  defaultLength: 3,  // the chip spans (sizeX + 1) * 32
  defaultFlags: FULL_ADDER_BITS,  // the interactive constructor sets it (FullAdderElm.java:25)
  defaults: { bits: 4, highVoltage: 5 },
  parse: (t, e, warn) => {
    // The bits token exists only under FLAG_BITS; a flagless line is the
    // 1-bit adder the file constructor defaults to, kept byte-exact on save.
    // The high-voltage token, when FLAG_CUSTOM_VOLTAGE says one follows, comes
    // after it (ChipElm.java:51-56).
    const flagged = (e.flags & FULL_ADDER_BITS) !== 0;
    chipCommonTokens(t, e, flagged, normalizeFullAdderBits, 'adder', warn);
    if (!flagged) e.params.bits = 1;
    chipBitOrderParam(e);
  },
  dump: (e) => {
    const out: (string | number)[] = [];
    if ((e.flags & FULL_ADDER_BITS) !== 0) out.push(e.params.bits ?? 4);
    const hv = e.params.highVoltage;
    if (hv !== undefined && hv !== 5) out.push(hv);
    return out;
  },
  dumpFlags: (e) => chipBitOrderFlags(e, chipDumpFlags(e)),
  fields: [
    { name: 'bits', label: '# of Bits', min: 1, max: 16, integer: true },
    { name: 'highVoltage', label: 'High logic voltage', unit: 'V' },
  ],
  draw: drawFullAdder,
};

File: engine/core/src/elements/full_adder.rs
//! Ripple-carry full adder (FullAdderElm.java, dump 196): `bits` A inputs and
//! `bits` B inputs summed with a carry-in into `bits` S outputs and a
//! carry-out. The width is the `bits` file token under FLAG_BITS, or 1 for a
//! flagless line, exactly the file constructor's default (FullAdderElm.java:
//! 30-35). Every output is one voltage source to ground through the shared
//! chip base; the sum recomputes combinationally each step.

use crate::element::{Base, Element, SimCtx};
use crate::elements::chip::{Chip, ChipPin};
use crate::spec::ElementSpec;
use crate::stamp::Stamper;

/// File flag saying a `bits` token follows the common chip fields; the
/// interactive constructor always sets it, so only a hand-edited line is
/// flagless (FullAdderElm.java:25, :36).
const FLAG_BITS: i64 = 2;

pub struct FullAdder {
    chip: Chip,
    bits: usize,
    /// True under upstream's BIT_ORDER_BUS (ChipElm.java:37): the A, B and S
    /// groups share one coordinate each, told apart by per-post bit tags.
    bus: bool,
}

impl FullAdder {
    pub fn new(spec: &ElementSpec) -> Self {
        // The width comes from the token under FLAG_BITS; a flagless line
        // means a 1-bit adder. The clamp keeps a corrupt file token from
        // allocating 3*bits + 2 posts.
        let bits = if spec.flag(FLAG_BITS) {
            (spec.param("bits", 4.0) as usize).clamp(1, 16)
        } else {
            1
        };
        let bus = spec.flag(crate::elements::chip::FLAG_BIT_ORDER_BUS);
        // Pin order (setupPins, FullAdderElm.java:41-56): the A and B inputs
        // on the west MSB first, the S outputs on the east, then the carry-in
        // and carry-out (FullAdderElm.java:50-54). No pin is saved to the file.
        let mut pins = Vec::with_capacity(3 * bits + 2);
        for _ in 0..bits {
            pins.push(ChipPin::input()); // A_i
        }
        for _ in 0..bits {
            pins.push(ChipPin::input()); // B_i
        }
        for _ in 0..bits {
            pins.push(ChipPin::output(false)); // S_i
        }
        pins.push(ChipPin::input()); // 3*bits+0 carry in
        pins.push(ChipPin::output(false)); // 3*bits+1 carry out
        Self {
            chip: Chip::new(spec, pins),
            bits,
            bus,
        }
    }

    fn execute(&mut self) {
        // A ripple carry walks the bits low to high (FullAdderElm.java:62-71).
        // The carry-in sits at 3*bits, the carry-out at 3*bits+1
        // (FullAdderElm.java:50-54).
        let mut carry = self.chip.pins[3 * self.bits].value as u32;
        for i in 0..self.bits {
            let sum =
                self.chip.pins[i].value as u32 + self.chip.pins[self.bits + i].value as u32 + carry;
            carry = u32::from(sum > 1);
            self.chip.write_output(2 * self.bits + i, sum & 1 == 1);
        }
        self.chip.write_output(3 * self.bits + 1, carry == 1);
    }
}

impl Element for FullAdder {
    fn kind(&self) -> &'static str {
        "fullAdder"
    }
    fn base(&self) -> &Base {
        &self.chip.base
    }
    fn base_mut(&mut self) -> &mut Base {
        &mut self.chip.base
    }
    fn post_count(&self) -> usize {
        3 * self.bits + 2
    }
    /// Bus mode: the A, B and S groups each collapse onto one coordinate, and
    /// makeBitPins runs them unreversed, so pin p carries bit `p` within its
    /// block (FullAdderElm.java:44-48); execute reads exactly that mapping
    /// (:62-64).
    fn post_bus_z(&self, post: usize) -> usize {
        if !self.bus {
            return 0;
        }
        if post < self.bits {
            post
        } else if post < 2 * self.bits {
            post - self.bits
        } else if post < 3 * self.bits {
            post - 2 * self.bits
        } else {
            0
        }
    }
    fn voltage_source_count(&self) -> usize {
        self.chip.voltage_source_count()
    }
    fn voltage_source_nodes(&self, k: usize) -> (usize, usize) {
        self.chip.voltage_source_nodes(k)
    }
    /// No current path between the posts: only the outputs' sources tie their
    /// nodes to ground (ChipElm.java:467).
    fn connects(&self, _a: usize, _b: usize) -> bool {
        false
    }

    fn stamp(&mut self, _ctx: &SimCtx, s: &mut Stamper) {
        self.chip.stamp(s);
    }

    fn start_iteration(&mut self, _ctx: &SimCtx) {
        if self.chip.read_inputs() {
            self.execute();
        }
        self.chip.commit_clock();
    }

    fn do_step(&mut self, _ctx: &SimCtx, s: &mut Stamper) {
        self.chip.do_step(s);
    }

    fn calculate_current(&mut self, _ctx: &SimCtx) {
        self.chip.base.current = 0.0;
    }

    fn current_into_node(&self, post: usize) -> f64 {
        self.chip.current_into_node(post)
    }

    fn set_param(&mut self, name: &str, value: f64) -> bool {
        match name {
            "highVoltage" => self.chip.high_voltage = value,
            // "bits" changes the post count, which only a full rebuild can
            // reallocate.
            _ => return false,
        }
        true
    }

    fn reset(&mut self) {
        self.chip.reset();
    }
}
